t2709: shared-gh-wrappers.sh: make REST-fallback source path resolution cross-shell (bash + zsh)

# t2709: shared-gh-wrappers.sh: make REST-fallback source path resolution cross-shell (bash + zsh)

`shared-gh-wrappers.sh:44-46` resolves its own directory via `${BASH_SOURCE[0]%/*}` and then sources `shared-gh-wrappers-rest-fallback.sh` relative to that directory. When sourced from a shell that does not populate `BASH_SOURCE[0]` (zsh, most notably), the parameter expansion returns empty and the source command tries to load `/shared-gh-wrappers-rest-fallback.sh` (leading-slash absolute path), emitting the warning:

```text
shared-gh-wrappers.sh:source:46: no such file or directory: /shared-gh-wrappers-rest-fallback.sh
```

The warning is non-fatal — the primary GraphQL path keeps working — but every single `source ~/.aidevops/agents/scripts/shared-constants.sh` invocation from a zsh session (including AI-agent Bash tools that run under zsh) emits it, polluting logs and confusing operators about whether the framework loaded correctly.

## Session Origin

Interactive session, surfaced during the t2708 fix (#20355 / PR #20356). The warning appeared on every `gh_create_issue` / `gh_create_pr` call in that worktree session.

## Why

- **The bash re-exec guard in `shared-constants.sh` only fires on execution, not sourcing.** When a user or agent runs `source ~/.aidevops/agents/scripts/shared-constants.sh` from zsh, the script is parsed by zsh. Zsh does not populate `BASH_SOURCE` — it has its own `funcfiletrace` array and the `${(%):-%x}` prompt-expansion idiom for the equivalent.
- **The current expansion `${BASH_SOURCE[0]%/*}` silently returns empty under zsh,** yielding the absolute path `/shared-gh-wrappers-rest-fallback.sh` which never exists. The t2574 REST fallback therefore isn't loaded in zsh-sourced sessions — a silent functional regression. If zsh-sourced wrappers hit GraphQL rate limits, they cannot fall back to REST.
- **macOS ships zsh as the default login shell** (since Catalina). Any user doing `source ~/.aidevops/agents/scripts/shared-constants.sh` from their default shell hits this.
- A clean fix here unblocks t2689/t2574's full REST-fallback coverage in any zsh-sourced path.

## What

Replace the self-directory calculation at line 44 with a shell-agnostic resolver that:

1. Uses `SCRIPT_DIR` if already set (the common case — `shared-constants.sh` sets this before sourcing us).
2. Falls back to `BASH_SOURCE[0]` when present (executed or sourced under bash).
3. Falls back to the zsh-specific `${(%):-%x}` when `ZSH_VERSION` is set.
4. Guards the `source` call behind a file-exists test so a missing fallback degrades silently instead of warning.

Verification: no warning on `zsh -c 'source ~/.aidevops/agents/scripts/shared-constants.sh'` and no warning on `bash -c 'source ~/.aidevops/agents/scripts/shared-constants.sh'`. In both cases the REST fallback helpers should be defined after sourcing (`type -t _gh_issue_create_rest` returns `function`).

## How

### Files Scope

- `.agents/scripts/shared-gh-wrappers.sh`
- `.agents/scripts/tests/test-shared-gh-wrappers-source.sh` (new)

### Implementation

Replace lines 44-46 of `shared-gh-wrappers.sh`.

**oldString:**

```bash
_SHARED_GH_WRAPPERS_DIR="${BASH_SOURCE[0]%/*}"
# shellcheck source=shared-gh-wrappers-rest-fallback.sh
source "${_SHARED_GH_WRAPPERS_DIR}/shared-gh-wrappers-rest-fallback.sh"
```

**newString:**

```bash
# Resolve own directory across bash (BASH_SOURCE) and zsh (%x prompt flag).
# Prefer SCRIPT_DIR when already set (the common path — shared-constants.sh
# sets it before sourcing us). Silently skip the REST fallback when the
# sibling file is not locatable; the primary GraphQL path keeps working.
_SHARED_GH_WRAPPERS_DIR="${SCRIPT_DIR:-}"
if [[ -z "$_SHARED_GH_WRAPPERS_DIR" ]]; then
	if [[ -n "${BASH_SOURCE[0]:-}" ]]; then
		_SHARED_GH_WRAPPERS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" 2>/dev/null && pwd)" || _SHARED_GH_WRAPPERS_DIR=""
	elif [[ -n "${ZSH_VERSION:-}" ]]; then
		# zsh: %x prompt flag expands to the current script path when sourced
		# shellcheck disable=SC2296
		_SHARED_GH_WRAPPERS_DIR="$(cd "$(dirname "${(%):-%x}")" 2>/dev/null && pwd)" || _SHARED_GH_WRAPPERS_DIR=""
	fi
fi
if [[ -n "$_SHARED_GH_WRAPPERS_DIR" && -f "$_SHARED_GH_WRAPPERS_DIR/shared-gh-wrappers-rest-fallback.sh" ]]; then
	# shellcheck source=shared-gh-wrappers-rest-fallback.sh
	source "$_SHARED_GH_WRAPPERS_DIR/shared-gh-wrappers-rest-fallback.sh"
fi
```

Notes:

- The `${(%):-%x}` construct is zsh's canonical "currently sourced file" — the `(%)` flag applies prompt-expansion to the default value `%x` (which is the prompt escape for "file of the executing script"). Bash never parses this path because it's guarded behind the `ZSH_VERSION` check.
- The `shellcheck disable=SC2296` is required because ShellCheck flags `(...)` parameter flags as unknown in bash mode. That's expected — the line is inside a zsh-only branch.
- The degraded-mode `:` path (neither `SCRIPT_DIR` nor `BASH_SOURCE` nor `ZSH_VERSION` — some exotic shell) silently declines to load the fallback. The primary GraphQL path still works; REST fallback simply won't trigger when GraphQL is exhausted in that shell.

### Tests

Create `.agents/scripts/tests/test-shared-gh-wrappers-source.sh` with two invocations:

```bash
# Bash path
bash -c 'source .agents/scripts/shared-constants.sh 2>&1' | grep -q 'shared-gh-wrappers-rest-fallback.sh' \
	&& fail "bash source emitted warning" || pass "bash sources cleanly"

# Zsh path (only if zsh is available)
if command -v zsh >/dev/null 2>&1; then
	zsh -c 'source .agents/scripts/shared-constants.sh 2>&1' | grep -q 'shared-gh-wrappers-rest-fallback.sh' \
		&& fail "zsh source emitted warning" || pass "zsh sources cleanly"
	# Also verify REST fallback helpers loaded under zsh
	zsh -c 'source .agents/scripts/shared-constants.sh && type _gh_issue_create_rest 2>/dev/null' | grep -q 'function' \
		|| fail "zsh did not load REST fallback helpers"
fi
```

The test should be skipped (not failed) if zsh is unavailable in CI.

## Acceptance

1. `zsh -c 'source .agents/scripts/shared-constants.sh 2>&1'` produces zero output about `shared-gh-wrappers-rest-fallback.sh`.
2. `bash -c 'source .agents/scripts/shared-constants.sh 2>&1'` still produces zero output (no regression).
3. Under zsh-sourced shared-constants.sh, `type _gh_issue_create_rest` reports a function (REST fallback loaded correctly).
4. New test `.agents/scripts/tests/test-shared-gh-wrappers-source.sh` passes.
5. No regression in existing tests (`test-gh-wrapper-rest-fallback.sh`, etc.).

## Files to modify

- **EDIT**: `.agents/scripts/shared-gh-wrappers.sh` lines 44-46 (replace the path-resolution + source block).
- **NEW**: `.agents/scripts/tests/test-shared-gh-wrappers-source.sh` — verify bash + zsh both source cleanly.

## Reference

- t2574 / GH#20255 — original REST fallback work that added this sourcing mechanism.
- t2689 / GH#20332 — extended REST fallback to read paths (same sourcing structure).
- t2688 / GH#20302 — `local -n` namerefs replaced for zsh/bash-3.2 compat; similar cross-shell compat work.
- Zsh `${(%):-%x}` idiom documented in `zshexpn(1)` "Parameter Expansion Flags" → `%` flag.

<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.8.91 plugin for [OpenCode](https://opencode.ai) v1.14.20 with claude-opus-4-7 spent 8h 32m and 534,896 tokens on this with the user in an interactive session.

<!-- ci-feedback:PR20375 -->
## CI Failure Feedback (from PR #20375)

The previous worker's PR #20375 had failing required CI checks. The PR has been
closed and this issue re-queued for dispatch. The next worker should address these failures.

### Failing checks

- **Complexity Analysis**: fail — [https://github.com/marcusquinn/aidevops/actions/runs/24747352587/job/72404007294](https://github.com/marcusquinn/aidevops/actions/runs/24747352587/job/72404007294)

### Worker guidance

1. Check out a fresh branch from `origin/main` (do NOT reuse the old branch)
2. Read the failing check URLs above for specific error messages
3. Fix the issues in the code, not in the CI config
4. Ensure all checks pass locally before pushing

_Routed by deterministic merge pass (pulse-merge.sh)._

--- a/.agents/scripts/shared-gh-wrappers.sh
+++ b/.agents/scripts/shared-gh-wrappers.sh
@@ -41,9 +41,22 @@ _SHARED_GH_WRAPPERS_LOADED=1
 # t2689: Extended to READ paths — _rest_issue_view, _rest_issue_list.
 # Provides _gh_should_fallback_to_rest, _gh_issue_{create,comment,edit}_rest,
 # _gh_pr_create_rest, _rest_issue_view, _rest_issue_list.
-_SHARED_GH_WRAPPERS_DIR="${BASH_SOURCE[0]%/*}"
-# shellcheck source=shared-gh-wrappers-rest-fallback.sh
-source "${_SHARED_GH_WRAPPERS_DIR}/shared-gh-wrappers-rest-fallback.sh"
+# Resolve own directory cross-shell (bash + zsh).
+# Priority: (1) BASH_SOURCE[0] under bash; (2) _SC_SELF set by shared-constants.sh
+# before sourcing us — both bash and zsh populate it correctly via ${0:-}, and it
+# points to shared-constants.sh which lives in the same directory as the REST
+# fallback; (3) absent both, silently skip — the primary GraphQL path still works.
+if [[ -n "${BASH_SOURCE[0]:-}" ]]; then
+	_SHARED_GH_WRAPPERS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" 2>/dev/null && pwd)" || _SHARED_GH_WRAPPERS_DIR=""
+elif [[ -n "${_SC_SELF:-}" ]]; then
+	_SHARED_GH_WRAPPERS_DIR="${_SC_SELF%/*}"
+else
+	_SHARED_GH_WRAPPERS_DIR=""
+fi
+if [[ -n "$_SHARED_GH_WRAPPERS_DIR" && -f "$_SHARED_GH_WRAPPERS_DIR/shared-gh-wrappers-rest-fallback.sh" ]]; then
+	# shellcheck source=shared-gh-wrappers-rest-fallback.sh
+	source "$_SHARED_GH_WRAPPERS_DIR/shared-gh-wrappers-rest-fallback.sh"
+fi
 
 # =============================================================================
 # GitHub Token Workflow Scope Check (t1540)

--- a/.agents/scripts/tests/test-shared-gh-wrappers-source.sh
+++ b/.agents/scripts/tests/test-shared-gh-wrappers-source.sh
@@ -1,0 +1,119 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
+#
+# test-shared-gh-wrappers-source.sh — t2709 / GH#20357 regression guard.
+#
+# Asserts that sourcing shared-constants.sh (which sources shared-gh-wrappers.sh)
+# emits zero output about shared-gh-wrappers-rest-fallback.sh under both bash
+# and zsh, and that the REST fallback helpers are loaded correctly in both.
+#
+# Root cause: shared-gh-wrappers.sh:44-46 used ${BASH_SOURCE[0]%/*} to resolve
+# its own directory. Under zsh, BASH_SOURCE is not populated, so the expansion
+# returned empty, and the subsequent source call tried to load
+# /shared-gh-wrappers-rest-fallback.sh (leading-slash absolute path), emitting:
+#   shared-gh-wrappers.sh:source:46: no such file or directory: /shared-gh-wrappers-rest-fallback.sh
+# Additionally the previous attempt used ${(%):-%x} (zsh-only syntax) which
+# caused shfmt to fail parsing the file, triggering the AWK fallback that
+# reported a false nesting depth of 14 and failed the CI regression gate.
+#
+# Fix: use _SC_SELF (set by shared-constants.sh before sourcing us) as the zsh
+# fallback — pure bash syntax, shfmt-parseable, same directory.
+#
+# Tests:
+#   1. bash: source shared-constants.sh emits zero REST-fallback warnings
+#   2. bash: _gh_issue_create_rest is defined after sourcing
+#   3. zsh:  source shared-constants.sh emits zero REST-fallback warnings (t2709)
+#   4. zsh:  _gh_issue_create_rest is defined after sourcing (t2709)
+
+set -uo pipefail
+
+SCRIPT_DIR_TEST="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)" || exit 1
+SCRIPTS_DIR="$(cd "${SCRIPT_DIR_TEST}/.." && pwd)" || exit 1
+REPO_ROOT="$(cd "${SCRIPTS_DIR}/../.." && pwd)" || exit 1
+
+if [[ -t 1 ]]; then
+	TEST_GREEN=$'\033[0;32m'
+	TEST_RED=$'\033[0;31m'
+	TEST_BLUE=$'\033[0;34m'
+	TEST_NC=$'\033[0m'
+else
+	TEST_GREEN="" TEST_RED="" TEST_BLUE="" TEST_NC=""
+fi
+
+TESTS_RUN=0
+TESTS_FAILED=0
+
+pass() {
+	TESTS_RUN=$((TESTS_RUN + 1))
+	printf '  %sPASS%s %s\n' "$TEST_GREEN" "$TEST_NC" "$1"
+	return 0
+}
+
+fail() {
+	TESTS_RUN=$((TESTS_RUN + 1))
+	TESTS_FAILED=$((TESTS_FAILED + 1))
+	printf '  %sFAIL%s %s\n' "$TEST_RED" "$TEST_NC" "$1"
+	if [[ -n "${2:-}" ]]; then
+		printf '       %s\n' "$2"
+	fi
+	return 0
+}
+
+skip() {
+	TESTS_RUN=$((TESTS_RUN + 1))
+	printf '  %sSKIP%s %s\n' "$TEST_BLUE" "$TEST_NC" "$1"
+	return 0
+}
+
+# =============================================================================
+# Test 1 + 2: bash — source emits no REST-fallback warning, helper defined
+# =============================================================================
+printf '\n=== bash source tests ===\n'
+
+bash_output=$(bash -c "source '${SCRIPTS_DIR}/shared-constants.sh' 2>&1" || true)
+if printf '%s\n' "$bash_output" | grep -q 'shared-gh-wrappers-rest-fallback.sh'; then
+	fail "bash source emits REST-fallback warning" "$bash_output"
+else
+	pass "bash source emits no REST-fallback warning"
+fi
+
+bash_fn=$(bash -c "source '${SCRIPTS_DIR}/shared-constants.sh' 2>/dev/null && type _gh_issue_create_rest 2>/dev/null" || true)
+if printf '%s\n' "$bash_fn" | grep -q 'function\|_gh_issue_create_rest'; then
+	pass "bash: _gh_issue_create_rest defined after source"
+else
+	fail "bash: _gh_issue_create_rest NOT defined after source" "$bash_fn"
+fi
+
+# =============================================================================
+# Test 3 + 4: zsh — source emits no REST-fallback warning, helper defined
+# =============================================================================
+printf '\n=== zsh source tests ===\n'
+
+if ! command -v zsh >/dev/null 2>&1; then
+	skip "zsh not available — skipping zsh tests (not a failure)"
+	skip "zsh: _gh_issue_create_rest defined after source (zsh unavailable)"
+else
+	zsh_output=$(zsh -c "source '${SCRIPTS_DIR}/shared-constants.sh' 2>&1" || true)
+	if printf '%s\n' "$zsh_output" | grep -q 'shared-gh-wrappers-rest-fallback.sh'; then
+		fail "zsh source emits REST-fallback warning (t2709)" "$zsh_output"
+	else
+		pass "zsh source emits no REST-fallback warning (t2709)"
+	fi
+
+	zsh_fn=$(zsh -c "source '${SCRIPTS_DIR}/shared-constants.sh' 2>/dev/null && type _gh_issue_create_rest 2>/dev/null" || true)
+	if printf '%s\n' "$zsh_fn" | grep -q 'function\|_gh_issue_create_rest\|shell function'; then
+		pass "zsh: _gh_issue_create_rest defined after source (REST fallback loaded)"
+	else
+		fail "zsh: _gh_issue_create_rest NOT defined after source" "$zsh_fn"
+	fi
+fi
+
+# =============================================================================
+# Summary
+# =============================================================================
+printf '\n=== Results: %d/%d passed ===\n' "$((TESTS_RUN - TESTS_FAILED))" "$TESTS_RUN"
+if [[ "$TESTS_FAILED" -gt 0 ]]; then
+	exit 1
+fi
+exit 0

